Conversion examples in the documentation and docstrings

h-mayorquin opened the pull request:
## Motivation

@bendichter mentioned in of our team meetings that it would be an idea to have a gallery of examples that we could quicly point out to the users. Also, we have issue #442 to add examples in the docstrings of non-data. This PR is to test whether it is feasible to integrate both approaches and at the same subject the docstring and the example gallery to testing suite using the capabilities provided by `doctest`:

To-do:
- [x] Add an example to the gallery to showcase format and idea.
- [x] Test locally with test.
- [x] Add tests to CI.

After the gallery examples are done we can add links to each of the docstrings in their respective data-interfaces so the examples are accessible from there.

bendichter commented on docs/conversion_examples_gallery/recording_and_sorting/neuroscope.rst:
```diff
@@ -0,0 +1,29 @@
+We use this for the metadata:
+    
+    >>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))    
+    >>> metadata = {"NWBFile":{"session_start_time":session_start_time}}
```
add a .get_metadata command

h-mayorquin replied:
Yes, I did not push this. Question:
Would you like to have the output of conversion.get_metadata() printed and explained or would you rather just have the call there and nothing else?

h-mayorquin replied:
See the new file for a comparison.

bendichter commented:
consider using [sphinx.ext.doctest](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html#directive-testcode)

h-mayorquin commented:
> consider using [sphinx.ext.doctest](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html#directive-testcode)

Yes, I am trying to find out if this can work as I expect that we can get prettier format that way. One problem with this `sphinx.ext.doctest` is that the tests are done at the time when the documentation is built.  This is problematic for us because the gin_test files are not available in the read the docs environment and my impression is that we have way less control over that one than what we have for the CI environment here in github. 

I am checking today if we can have `sphinx.ext.doctest` format but doing the testing with pytest in github CI.

bendichter commented:
@h-mayorquin makes sense. I would avoid downloading data in the docs build because I think we'll run up against space and time contraints on readthedocs

h-mayorquin commented:
OK, so now we have a pipe-line going on here. I added testing in CI and now I added some utilities so the code can be display with and without the prompt, I also added copying functionality.

Right now, I have two format proposals in the neuroscope docpage:
Could we chose the format, accept this as a template and I will all the other interfaces based on this?
I am taking a long flight tomorrow and I would be good to work on that on the flight.

bendichter commented on docs/conversion_examples_gallery/recording_and_sorting/neuroscope.rst:
```diff
@@ -0,0 +1,57 @@
+    >>> # Creates a datetime object with date the first of Junuary of 2020 in the US/Pacific time-zone
+    >>> metadata = interface.get_metadata()
+    >>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))
+    >>> metadata["NWBFile"] = dict(session_start_time=session_start_time)
```
```suggestion
    >>> metadata["NWBFile"].update(session_start_time=session_start_time)
```

I prefer this syntax. The way you have it here, any fields in NWBFile other than `session_start_time` would be removed. You could also do `metadata["NWBFile"]["session_start_time"] = session_start_time`, but I prefer doing the dictionary update because you can assign multiple fields at once if you want to, e.g.

```python
metadata["NWBFile"].update(session_start_time=session_start_time, notes="hello")
```

h-mayorquin replied:
Right, this makes more sense. 
Done in the latest commit.

h-mayorquin replied:
@bendichter 
I just realized though, that this fails because the metadata does not necessarily have the key `NWBFile` at that point in time.

bendichter commented on docs/conf.py:
```diff
@@ -7,7 +7,16 @@
-extensions = ["sphinx.ext.autodoc", "sphinx.ext.autosummary", "sphinx.ext.napoleon", "myst_parser"]
+extensions = [
+    "sphinx.ext.autodoc",
+    "sphinx.ext.doctest",
```
are we still using this?

h-mayorquin replied:
I removed it now.

bendichter commented on docs/conversion_examples_gallery/recording_and_sorting/neuroscope.rst:
```diff
@@ -0,0 +1,57 @@
+Then we creates a datetime object representing the date of the first of January of 2020 in the US/Pacific time-zone. 
+We use this for the metadata:
+    
+    >>> # Creates a datetime object with date the first of Junuary of 2020 in the US/Pacific time-zone
+    >>> metadata = interface.get_metadata()
+    >>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))
```
```suggestion
    >>> # Extract what metadata we can from the source files
    >>> metadata = interface.get_metadata()
    >>>
    >>> # session_start_time is required for conversion. If it cannot be inferred 
    >>> # automatically from the source files you must supply one.
    >>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))
```

h-mayorquin replied:
Done in the latest commit.

bendichter commented on docs/conversion_examples_gallery/recording_and_sorting/neuroscope.rst:
```diff
@@ -0,0 +1,57 @@
+    >>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))
+    >>> metadata["NWBFile"] = dict(session_start_time=session_start_time)
+    >>>
+    >>> # Chose a path for saving the nwb file and run the conversion
```
```suggestion
    >>> # Choose a path for saving the nwb file and run the conversion
```

h-mayorquin replied:
Done.

bendichter commented on docs/conversion_examples_gallery/recording_and_sorting/neuroscope.rst:
```diff
@@ -0,0 +1,60 @@
+Neuroscope data conversion
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Example of how to convert ``Neuroscope`` to nwb:
```
```suggestion
Convert Neuroscope data to NWB using :py:class:`~nwb_conversion_tools.datainterfaces.NeuroscopeRecordingInterface`.
```

bendichter commented:
@h-mayorquin very cool! The tests and the toggling of the prompt look great!

weiglszonja commented on docs/conversion_examples_gallery/recording_and_sorting/neuroscope.rst:
```diff
@@ -0,0 +1,60 @@
+    >>> save_path = "./nwb_neuroscope2.nwb"
+    >>> interface.run_conversion(save_path=save_path, metadata=metadata)
+    >>>
+    >>> # If everything went well this should evaluate to True as the file was created
```
```suggestion
    >>> # If the conversion was successful this should evaluate to ``True`` as the file was created.
```

h-mayorquin replied:
Done! Thanks!

h-mayorquin commented on docs/conversion_examples_gallery/recording_and_sorting/neuroscope.rst:
```diff
@@ -0,0 +1,57 @@
+    True
+
+The other type of display:
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I added the changes and now this is the only one that remains. Which scrollbar, @weiglszonja? (maybe it disappear now but I did not know what you were talking about).

h-mayorquin commented on docs/conversion_example_gallery.rst:
```diff
@@ -0,0 +1,17 @@
+Examples of conversion
+======================
+
+The examples here are guaranteed to be running correctly with our testing suite. The aim to be a showcase of the library
+capabilities and their aim is to get you to hit ground running with your conversion.  
```
Thanks, now the we chose the the second format, this text is not present.

h-mayorquin replied:
Done! Thanks!

bendichter commented on docs/conversion_examples_gallery/recording/neuroscope.rst:
```diff
@@ -0,0 +1,31 @@
+    >>> # session_start_time is required for conversion. If it cannot be inferred 
+    >>> # automatically from the source files you must supply one.
+    >>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))
+    >>> metadata["NWBFile"] = dict(session_start_time=session_start_time)
```
```suggestion
    >>> metadata["NWBFile"].update(session_start_time=session_start_time)
```

h-mayorquin replied:
This will fail because in this example the metadata does not have the key `NWBFile` already.

bendichter replied:
ah

h-mayorquin commented:
Using fixtures from `pytest` together with [conftest configuration file](https://github.com/catalystneuro/nwb-conversion-tools/blob/d695a2ee66daa6b79c47ca91fdcf653c0dd1235a/docs/conversion_examples_gallery/conftest.py) we can define the `data_path` for the `doctests` in the same way that we do it for the gin tests. With that addition, the previously added action to the CI is no longer needed. The reason being that now the doctests can be run with a simple pytest command (for this, also some modifications to the `pyproject.toml` were [introduced](https://github.com/catalystneuro/nwb-conversion-tools/blob/d695a2ee66daa6b79c47ca91fdcf653c0dd1235a/pyproject.toml)). 

The advantage of this is that now the entire test suite -including the `doctests`- are run with one command, within the same action in the CI, and more importantly can be run locally without needing to modify the paths of the file for quicker debugging.

bendichter commented on docs/conversion_examples_gallery/sorting/phy.rst:
```diff
@@ -0,0 +1,28 @@
+    tzinfo = tz.gettz("US/Pacific")
+    session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))
+    session_start_time = session_start_time.replace(tzinfo=tzinfo).isoformat()
+    metadata["NWBFile"] = dict(session_start_time=session_start_time)
```
```suggestion
    metadata["NWBFile"].update(session_start_time=session_start_time)
```

bendichter commented on docs/conversion_examples_gallery/recording/intan.rst:
```diff
@@ -0,0 +1,32 @@
+    >>> # session_start_time is required for conversion. If it cannot be inferred 
+    >>> # automatically from the source files you must supply one.
+    >>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific"))
+    >>> metadata["NWBFile"] = dict(session_start_time=session_start_time)
```
```suggestion
    >>> metadata["NWBFile"].update(session_start_time=session_start_time)
```

can you please try to do it this way?

bendichter replied:
Does intan already contain the datetime in the header?

CodyCBakerPhD replied:
Intan headers have a wealth of information, particularly about electrode metadata. But I don't think we extract much of it automatically just yet

h-mayorquin replied:
Hey, @bendichter, are you reading my comments above that this does not work when the metadata does not contain the key "NWBFile", already? 

Maybe I am missing something?

bendichter replied:
yeah I missed that earlier comment. OK, I see, this is correct. After we get this through let's look at ways we can pull this metadata from the source files.

bendichter commented on docs/conversion_examples_gallery/sorting/phy.rst:
```diff
@@ -0,0 +1,27 @@
+    >>> interface = PhySortingInterface(folder_path=folder_path)
+    >>> 
+    >>> metadata = interface.get_metadata()
+    >>> tzinfo = tz.gettz("US/Pacific")
```
```suggestion
```

bendichter commented on docs/conversion_examples_gallery/sorting/kilo.rst:
```diff
@@ -0,0 +1,27 @@
+    >>> interface = KiloSortingInterface(folder_path=folder_path)
+    >>> 
+    >>> metadata = interface.get_metadata()
+    >>> tzinfo = tz.gettz("US/Pacific")
```
```suggestion
```

bendichter commented on docs/conversion_examples_gallery/recording/spikegadgets.rst:
```diff
@@ -0,0 +1,31 @@
+    >>> # Extract what metadata we can from the source files
+    >>> metadata = interface.get_metadata()
+    >>> # For data provenance we add the time zone information to the conversion
+    >>> tzinfo = tz.gettz("US/Pacific")
```
```suggestion
```

bendichter commented:
@h-mayorquin this looks great! I say we merge this soon so that the tutorials we have are available to users and then work on the remaining formats.

bendichter commented:
@h-mayorquin the interfaces, e.g. `BlackrockRecordingExtractorInterface`, should have active links to the API, but the links are not working

bendichter commented on docs/conversion_examples_gallery/recording/openephys.rst:
```diff
@@ -0,0 +1,28 @@
+Open Ephys data conversion
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Convert Open Ehpys data to NWB using :py:class:`~nwb_conversion_tools.datainterfaces.OpenEphysRecordingExtractorInterface`.
```
```suggestion
Convert OpenEphys data to NWB using :py:class:`~nwb_conversion_tools.datainterfaces.OpenEphysRecordingExtractorInterface`.
```

h-mayorquin commented:
> BlackrockRecordingExtractorInterface

Is there any place (outside of these changes here) where this functionality is working?

bendichter commented:
@h-mayorquin you need the full path for the links to work, not the abbreviated paths. For instance, instead of `nwb_conversion_tools.datainterfaces.PhySortingDataInterface`, you need `nwb_conversion_tools.datainterfaces.ecephys.phy.phydatainterface.PhySortingDataInterface`. This makes the tilde notation all the more useful

h-mayorquin commented:
> @h-mayorquin you need the full path for the links to work, not the abbreviated paths. For instance, instead of `nwb_conversion_tools.datainterfaces.PhySortingDataInterface`, you need `nwb_conversion_tools.datainterfaces.ecephys.phy.phydatainterface.PhySortingDataInterface`. This makes the tilde notation all the more useful

Right, and some of them were not available in the autodocs/API. All of them should be working in the last commit.

h-mayorquin commented:
> I just used this type of setup to copy/paste some instructions for someone on the helpdesk, here's some feedback
> 
> (a) when I use the copy/paste 'button', it includes the annoying '>>>' characters which I then have to either remove or manually select the code block to paste without that
> 
> (b) when manually selecting and copy/pasting, this also grabs any empty lines as just the '>>>' characters, which is kind of annoying, I wish it would just copy/paste a blank newline

Thanks, @CodyCBakerPhD !
Which helpdesk? : O

I just tried this again here in docs generated for this PR and the two types of copying (button and manual) work in the non-annoying way for me. I tried in browser, vscode and gedit. Could be a windows thing? no ideas here.

CodyCBakerPhD commented:
> Which helpdesk? : O

Technically both, but mainly this one: https://github.com/NeurodataWithoutBorders/helpdesk/discussions/29


> I just tried this again here in docs generated for this PR and the two types of copying (button and manual) work in the non-annoying way for me. I tried in browser, vscode and gedit. Could be a windows thing? no ideas here.

Anything is possible with cross-platform stuff.

Here is what my screen looks like, windows 10 + google chrome (latest)

![image](https://user-images.githubusercontent.com/51133164/167476890-a3de46ad-95fe-453c-b073-3af4e735aeb3.png)

and output paste into GitHub window (#1 way this type of thing will be used) from hitting copy button

```python
>>> from datetime import datetime
>>> from dateutil import tz
>>> from pathlib import Path
>>> from nwb_conversion_tools import SpikeGadgetsRecordingInterface
>>>
>>> # For this interface we need to pass the location of the ``.bin`` file
>>> file_path = f"{ECEPHY_DATA_PATH}/spikegadgets/20210225_em8_minirec2_ac.rec"
>>> # Change the file_path to the location in your system
>>> interface = SpikeGadgetsRecordingInterface(file_path=file_path)
>>>
>>> # Extract what metadata we can from the source files
>>> metadata = interface.get_metadata()
>>> # For data provenance we add the time zone information to the conversion
>>> session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=tz.gettz("US/Pacific")).isoformat()
>>> metadata["NWBFile"] = dict(session_start_time=session_start_time)
>>>  # Choose a path for saving the nwb file and run the conversion
>>> save_path = f"{path_to_save_nwbfile}"
>>> interface.run_conversion(save_path=save_path, metadata=metadata)
>>>
>>> # If the conversion was successful this should evaluate to ``True`` as the file was created.
>>> Path(save_path).is_file()
True
```

h-mayorquin commented:
Maybe silly, but I was referring to copying after clicking on the `>>>` symbol (top-right) which should make the prompt symbols disappear and the copy then should work in the expected fashion. Did you try that?

h-mayorquin commented:
> Ahh, yes there is a '>>>' button next to the copy/paste button
> 
> ![image](https://user-images.githubusercontent.com/51133164/167479029-b4c1a0cc-adf3-4ca1-9ae2-a9fba707670b.png)
> 
> This was not that visually apparent on first use, though, it looked like just another/misaligned code block. Is there an advantage to having that button on 'enable' mode by default? As I recall the reason you needed the '>>>' lines in the first place was to get it to run the code in tests, but if there's a way to hide them by default in the actual rendering that would be nicer.

Yes, @bendichter  is ahead of you:
https://github.com/jurasofish/sphinx-toggleprompt/issues/11#issuecomment-1121084804

Sadly, not there yet. Maybe we can leave a boiler plate reminder in each of the pages to remind users that they can get rid of the prompt symbols.

CodyCBakerPhD commented:
> Yes, @bendichter is ahead of you:
https://github.com/jurasofish/sphinx-toggleprompt/issues/11#issuecomment-1121084804

> Sadly, not there yet. Maybe we can leave a boiler plate reminder in each of the pages to remind users that they can get rid of the prompt symbols.

Awesome, thanks! Always good to see active devs on top of such things

bendichter commented on docs/conversion_example_gallery.rst:
```diff
@@ -0,0 +1,31 @@
+    Blackrock <conversion_examples_gallery/recording/blackrock>
+    Intan <conversion_examples_gallery/recording/intan>
+    Neuroscope <conversion_examples_gallery/recording/neuroscope>
+    Open Ephys <conversion_examples_gallery/recording/openephys>
```
```suggestion
    OpenEphys <conversion_examples_gallery/recording/openephys>
```

bendichter commented on docs/conversion_example_gallery.rst:
```diff
@@ -0,0 +1,31 @@
+.. toctree::
+    :maxdepth: 1
+
+    Kilosorting <conversion_examples_gallery/sorting/kilo>
```
```suggestion
    Kilosort <conversion_examples_gallery/sorting/kilo>
```

bendichter commented on docs/conversion_examples_gallery/sorting/kilo.rst:
```diff
@@ -0,0 +1,26 @@
+Kilosorting data conversion
```
```suggestion
Kilosort data conversion
```

bendichter commented on docs/conversion_examples_gallery/sorting/kilo.rst:
```diff
@@ -0,0 +1,26 @@
+Kilosorting data conversion
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Convert Kilosorting data to NWB using :py:class:`~.nwb_conversion_tools.datainterfaces.ecephys.kilo.kilodatainterface.KiloSortingInterface`.
```
```suggestion
Convert Kilosort data to NWB using :py:class:`~.nwb_conversion_tools.datainterfaces.ecephys.kilo.kilodatainterface.KiloSortingInterface`.
```

CodyCBakerPhD commented:
(I think Axona also technically has a sorting interface, but it was more ad hoc as I recall so maybe skip for now?)

bendichter commented on docs/conversion_example_gallery.rst:
```diff
@@ -0,0 +1,32 @@
+Conversion Gallery
```
I'd put this file inside the conversion_examples_gallery directory. Then all of the reference can shorten, e.g. conversion_examples_gallery/recording/axona -> recording/axona

h-mayorquin replied:
I was thinking on having everything that is on `index.rst` (the top-level file) also in docs. The idea was the file structure would reflect the document structure.

h-mayorquin commented:
> Axona, blackrock, spikegadgets, spikeglx don't have spacing between comments consistent with the other interfaces (which looks slightly better)
> 
> OpenEphys doesn't have the same comment structure overall
> 
> Neuroscope and cellexplorer sorting examples?

Thanks, I align the comment structure for the former group.

OpenEphys produces output by default (click on the `>>>` to see what's need for testing).

Neuroscope and cellexplorer. Those ones are missing.
